[Standardization] K8s cluster robustness features (#414) (#361)
SovereignCloudStack/standards: Standards/scs-0215-v1-robustness-features.md

---
title: Robustness features for K8s clusters
type: Standard
status: Draft
track: KaaS
---

## Introduction

Kubernetes clusters in a productive environment are under the assumption to always perform perfectly without any major
interruptions. But due to external or unforeseen influences, clusters can be disrupted in their normal workflow, which
could lead to slow responsiveness or even malfunctions.
In order to possibly mitigate some problems for the Kubernetes clusters, robustness features should be introduced into
the SCS standards. These would harden the cluster infrastructure against several problems, making failures less likely.

## Motivation

A typical productive Kubernetes cluster could be hardened in many different ways, whereas probably many of these actions
would overlap and target similar weaknesses of a cluster.
For this version of the standard, the following points should be addressed:

* Kube-API rate limiting
* etcd compaction/defragmentation
* etcd backup

> * CA expiration avoidance

Reviewer: What does "CS" means?

Author: CA = Certificate Authority

These robustness features should mainly increase the up-time of the Kubernetes cluster by avoiding downtimes either
because of internal problems or external threads like "Denial of Service" attacks.

> Additionally, the ETCD database should be strengthened with these features in order to provide a secure and robust

Reviewer: In line 23 and 24 "etcd" is written in lower case and here in upper We should be consistent throughout the whole document.

Author: Made it consistent across the document.

### CA expiration avoidance

In order to secure the communication of a Kubernetes cluster, (TLS) certificates signed by a controlled
Certificate Authority (CA) can be used.
Normally, these certificates expire after a set period of time. In order to avoid expiration and failure of a cluster,
these certificates need to be rotated regularly and at best automatically.
It is important to either set `--rotate-server-certificates` as a command line parameter or set `rotateCertificates: true`
in the kubelet config or the `kubeletExtraArgs` of the cluster-template.yaml file. This activates the rotation of the
kubelet server certificate. Another recommendation is to set `serverTLSBootstrap: true`, which also enables the request
and rotation of the certificate for the kubelet according to the documentation.

A clusters certificates can either be rotated by updating the cluster, which according to the Kubernetes documentation
automatically renews the certificates, or by running the kubeadm certs renew command for the certificates that need to
updated, which can be seen in the following inline code

```bash
kubeadm certs renew all
```

> Since clusters conformant with the SCS standards would probably be updated within a 14 month time period, this

Reviewer: I would refer and link the standard, which dictates an update period of 14 months, here.

Author: Referred to the standard with a link here, but the referred standard isn't in the main branch yet. I nevertheless made the reference link use the main branch, since the standard will be merged soon (hopefully).

rotation can probably be assumed to happen. Nevertheless, the alternative can still be mentioned in the standard.
Additionally, the CSR needs to be approved manually due to security reasons with the commands

```bash
kubectl get csr
kubectl certificate approve <CSR>
```

> Another option to approve the CSRs would be to use a third-party controller that automates the process. One example for

Reviewer: Please do not use abbreviations. without introducing them. Additionally, I recommend to add a glossary at the beginning to the document, like in [scs-0302-v1-domain-manager-role.md](https://github.com/SovereignCloudStack/standards/blob/main/Standards/scs-0302-v1-domain-manager-role.md)

Author: Added a glossary and put the abbreviations at the first occurrence of their respective meanings.

under [Manual rotation of ca certificate](https://kubernetes.io/docs/tasks/tls/manual-rotation-of-ca-certificates/).

Further information can be found in the Kubernetes documentation:
[Kubeadm certs](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/)
[Kubelete TLS bootstrapping](https://kubernetes.io/docs/reference/access-authn-authz/kubelet-tls-bootstrapping/)

## Decision

Robustness features combine multiple aspects of increasing the security, hardness and
longevity of a Kubernetes cluster. The decisions will be separated into their respective
areas.

### Kube-API rate limiting

The number of requests send to the kube-api or Kubernetes API server SHOULD be limited
in order to protect the server against outages, deceleration or malfunctions due to an
overload of requests.
In order to do so, at least the following parameters SHOULD be set on a Kubernetes cluster:

* max-requests-inflight
* max-mutating-requests-inflight
* min-request-timeout

Values for these flags/parameters SHOULD be adapted to the needs of the environment and
the expected load.

> A cluster MUST also activate and configure a Ratelimit admission controller.

Reviewer: Why is limitation of number of requests a SHOULD but Ratelimit admission controller a MUST. Isn't Ratelimit admission controller a special case for the limitation of request as Kube API? If so, we should set number of request limitation to MUST. Otherwise line 161ff and link 173ff are contradictory.

Author: Set both lines to MUST.

This requires an `EventRateLimit` resource to be deployed on the Kubernetes cluster.
The following settings are RECOMMENDED for a cluster-wide deployment, but more
fine-grained rate limiting can also be applied, if this is necessary.

```yaml
kind: Configuration
apiVersion: eventratelimit.admission.k8s.io/v1alpha1
limits:
- burst: 20000
  qps: 5000
  type: Server
```

It is also RECOMMENDED to activate the Kubernetes API priority and fairness feature,
which also uses the aforementioned cluster parameters to better queue, schedule and
prioritize incoming requests.

### etcd compaction

> etcd needs to be cleaned up regularly, so that it functions correctly and doesn't take

Reviewer: I guess, etcd cleanup is a must and we should phrase it like that.

```suggestion
etcd must to be cleaned up regularly, so that it functions correctly and doesn't take
```

up too much space, which happens because of its increase of the keyspace.

To compact the etcd keyspace, the following flags/parameters MUST be set for etcd:

* auto-compaction-mode = periodic
* auto-compaction-retention = 8h

### etcd backup

> An etcd cluster MUST be backed up regularly. It is RECOMMENDED to adapt

Reviewer: etcd MUST be back up regularly, I agree. However, this requirement is senseless, without defining a backup strategy. IMO it is a huge difference in robustness, if a provider backups monthly or daily. I recommend to add a minimal backup cycle in this standard, a customer can expect for k8s cluster, that are SCS certified.

Author: There is a recommendation for the backup strategy after that which refers to the previous section.
I wouldn't further define (or make this subsection a MUST), since not all providers can do a backup (or multiple backups) every day.
I would maybe put in a weekly backup requirement, but more would be too fine-grained and defined for some providers.

10 minutes for the last 120 minutes, then one hourly for 1 day, then one daily for 2 weeks,
then one weekly for 3 months, then one monthly for 2 years, and after that a yearly backup.
At the very least, a backup MUST be done once a week.
These numbers can be adapted to the security setup and concerns like storage or network
usage. It is also RECOMMENDED to encrypt the backups in order to secure them further.
How this is done is up to the operator.

### CA expiration avoidance

It should be avoided, that certificates expire either on the whole cluster or for single components.
To avoid this scenario, certificates SHOULD be rotated regularly; in the
case of SCS, we REQUIRE at least a yearly certificate rotation.
To achieve a complete certificate rotation, the parameters `serverTLSBootstrap` and `rotateCertificates` MUST be set.

The certificates can be rotated by either updating the Kubernetes cluster, which automatically
renews certificates, or by manually renewing them with the command

```bash
kubeadm certs renew all
```

After this, new CSRs MUST be approved manually, normally done with

```bash
kubectl get csr
kubectl certificate approve <CSR>
```

> or be approved with a third-party controller, e.g. the [kubelet-csr-approver](https://github.com/postfinance/kubelet-csr-approver).

Author: I changed a few sentences in the document and especially in the `Decision` section in order to better reflect a few of the concerns you had. You're right that `kubeadm` was a bit to specific, so I only mentioned it as an example in the `Design Considerations`.

About the CSR part, I'm not really sure. Yes they don't HAVE to be used, but https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#kubelet-serving-certs reads like its a recommended way, especially if some external tools should be used with the cluster (like monitoring solutions).
But I'm really not an expert on this.

> or be approved with a third-party controller, e.g. the [kubelet-csr-approver](https://github.com/postfinance/kubelet-csr-approver).

Author: @joshmue Is this fine for now or are there any other problems with this?

> or be approved with a third-party controller, e.g. the [kubelet-csr-approver](https://github.com/postfinance/kubelet-csr-approver).

Reviewer 2: > You're right that kubeadm was a bit to specific, so I only mentioned it as an example in the Design Considerations.

:+1: 

> About the CSR part, I'm not really sure. [...] But I'm really not an expert on this.

I did not dig into that to much right now, so I'm not totally sure whether this should be mandated here, either (I'd tend not to).
If this is to be included and merged, someone should be sure about that, though.

> Yes they don't HAVE to be used, but https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#kubelet-serving-certs reads like its a recommended way, especially if some external tools should be used with the cluster (like monitoring solutions).

Again, that is kubeadm-centric documentation, so not necessarily relevant for other distributions.

Other than that, the current status has some SHOULD/can/MUST ambiguity problems:

> To avoid this scenario, certificates SHOULD be rotated regularly; in the
> case of SCS, we REQUIRE at least a yearly certificate rotation.

That could be understood two ways: Either someone could decide to be non-compliant with that SHOULD, hence making the REQUIRE-ment ineffective (That again would mandate *how* something is done, while the *what* is optional - pretty much the opposite of what standards usually do). Or someone could read the REQUIRE as actual requirement, making the SHOULD an effective MUST.

> The certificates can be rotated by either updating the Kubernetes cluster, which automatically
> renews certificates, or by manually renewing them. How this is done is dependent on the used K8s cluster.
> 
> After this, new CSRs MUST be approved manually or with a third-party controller, e.g. the [kubelet-csr-approver](https://github.com/postfinance/kubelet-csr-approver).

Kind of the same problem as above.

> It is also RECOMMENDED to renew the CA regularly to avoid an expiration of the CA.

:+1: 

> This standard doesn't set a timeline for this, since it is dependent on the CA.

Is this intended to say that the CA may be out of the cluster operator's control? That's not really clear.

> or be approved with a third-party controller, e.g. the [kubelet-csr-approver](https://github.com/postfinance/kubelet-csr-approver).

Author: > > Yes they don't HAVE to be used, but https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#kubelet-serving-certs reads like its a recommended way, especially if some external tools should be used with the cluster (like monitoring solutions).
>
> Again, that is kubeadm-centric documentation, so not necessarily relevant for other distributions.

Yes this documentation uses `kubeadm`, but I would argue that the CSR part is not-centered especially on `kubeadm` and can be applied to other K8s cluster "types" aswell.

> > To avoid this scenario, certificates SHOULD be rotated regularly; in the
> > case of SCS, we REQUIRE at least a yearly certificate rotation.
>
> That could be understood two ways: Either someone could decide to be non-compliant with that SHOULD, 
> hence making the REQUIRE-ment ineffective (That again would mandate *how* something is done, while the 
> *what* is optional - pretty much the opposite of what standards usually do). Or someone could read the 
> REQUIRE as actual requirement, making the SHOULD an effective MUST.
>
> > The certificates can be rotated by either updating the Kubernetes cluster, which automatically
> > renews certificates, or by manually renewing them. How this is done is dependent on the used K8s cluster.
> > 
> > After this, new CSRs MUST be approved manually or with a third-party controller, e.g. the [kubelet-csr-approver](https://github.com/postfinance/kubelet-csr-approver).
>
> Kind of the same problem as above.

I agree with the first part (and I fixed that), but the second part doesn't have the same problem.
The whole part is about certificate rotation (meaning its continuous), so the part with the `can` is refering to the certificate rotation part, which is written as a `MUST` now (`SHOULD` before, so that would be contradictive).
Also `can` isn't an RFC keyword (https://www.rfc-editor.org/rfc/rfc2119), so it's not defining anything, just explaining a process from this.

> or be approved with a third-party controller, e.g. the [kubelet-csr-approver](https://github.com/postfinance/kubelet-csr-approver).

Author: Updated with https://github.com/SovereignCloudStack/standards/pull/361/commits/84fcef3ff4670d82ac28ab5cbf85caf515678a01

> or be approved with a third-party controller, e.g. the [kubelet-csr-approver](https://github.com/postfinance/kubelet-csr-approver).

Reviewer 2: > > > Yes they don't HAVE to be used, but https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#kubelet-serving-certs reads like its a recommended way, especially if some external tools should be used with the cluster (like monitoring solutions).
> >
> > Again, that is kubeadm-centric documentation, so not necessarily relevant for other distributions.
> 
> Yes this documentation uses `kubeadm`, but I would argue that the CSR part is not-centered especially on `kubeadm` and can be applied to other K8s cluster "types" aswell.

I did not really use `kubeadm` all that much for a few years now. But looking at its [docs](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#renew-certificates-with-the-kubernetes-certificates-api) (mind the `Caution` paragraph), handling certificates this way seems to be the exception even for `kubeadm` setups, nevermind other distributions.

> I agree with the first part (and I fixed that)

:+1: 

> The whole part is about certificate rotation (meaning its continuous), so the part with the can is refering to the certificate rotation part, which is written as a MUST now (SHOULD before, so that would be contradictive).
> Also can isn't an RFC keyword (https://www.rfc-editor.org/rfc/rfc2119), so it's not defining anything, just explaining a process from this.

Still, this MUST refers to `how` certificates should be approved. In some setups, there probably is not even a concept for approval present.

And if there is, it is buried so low that it is unpractical to check for.
For example: Could [k3s certificate rotation](https://docs.k3s.io/cli/certificate) ever be compliant to this standard's MUST? That is cumbersome to (dis-)prove and does not add much value over the simple requirement of "Certificates should be rotated regularly".

> or be approved with a third-party controller, e.g. the [kubelet-csr-approver](https://github.com/postfinance/kubelet-csr-approver).

Author: > > The whole part is about certificate rotation (meaning its continuous), so the part with the can is refering to the > > certificate rotation part, which is written as a MUST now (SHOULD before, so that would be contradictive).
> >Also can isn't an RFC keyword (https://www.rfc-editor.org/rfc/rfc2119), so it's not defining anything, just explaining a process from this.
>
> Still, this MUST refers to how certificates should be approved. In some setups, there probably is not even a
> concept for approval present.
>
> And if there is, it is buried so low that it is unpractical to check for.
> For example: Could [k3s certificate rotation](https://docs.k3s.io/cli/certificate) ever be compliant to this 
> standard's MUST? That is cumbersome to (dis-)prove and does not add much value over the simple requirement
> of "Certificates should be rotated regularly".

Removed the sentence in question.

kubelet server certificate. Another recommendation is to set `serverTLSBootstrap: true`, which also enables the request
and rotation of the certificate for the kubelet by requesting them from `certificates.k8s.io`
according to the documentation under [Kubeadm certs].

A clusters certificates can be rotated by updating the cluster, which according to the Kubernetes documentation
automatically renews the certificates.
Another method would be the manual renewal, which can be done through multiple methods, depending on the K8s cluster
used. An example for a K8s cluster set up with `kubeadm` would be to execute the command

```bash
kubeadm certs renew all
```

Since clusters conformant with the SCS standards would probably be updated within the time period described in the
standard [SCS-0210-v2](https://github.com/SovereignCloudStack/standards/tree/main/Standards/scs-0210-v2-k8s-version-policy.md),
this rotation can probably be assumed to happen, because of the cluster update functionality.

It is also important to mention, that due to security reasons, the CSR needs to be approved manually with the commands

```bash
kubectl get csr
kubectl certificate approve <CSR>
```

> in order to complete a certificate rotation.

Reviewer 2: Related to https://github.com/SovereignCloudStack/standards/pull/361#discussion_r1448822622

> It is also important to mention, that due to security reasons, the CSR needs to be approved manually with the commands

Sure? https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#manual-certificate-renewal does not seem to mention it.

Author: Rewrote the section to hopefully fit better.

Author: Summoning @joshmue again to resolve (and maybe approve)

Reviewer 2: The rewording from hard "needs to" etc. to "may need to" reduces the risk of strictly false statements, indeed.

My point (from https://github.com/SovereignCloudStack/standards/pull/361#discussion_r1447407033) still is that information should be verified with a high degree of certainty, which I do not feel there is:

> > About the CSR part, I'm not really sure. [...] But I'm really not an expert on this.
> >
> I did not dig into that to much right now, so I'm not totally sure whether this should be mandated here, either (I'd tend not to).
> If this is to be included and merged, someone should be sure about that, though.

Author: So I read up on most of the stuff again (https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs, https://kubernetes.io/docs/reference/access-authn-authz/kubelet-tls-bootstrapping, https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/) and came to the conclusion, to rewrite the `Certificate rotation` section, since there were quite a few errors and mixups in this part. IMO its now better structured and correct.

I also shortened the section related to this under `Decision` to the relevant stuff.

Reviewer 2: Looks way better to me, as well! :+1:

|-----------------------------|----------------------------------------------------------------|
| Certificate Authority       | Trusted organization that issues digital certificates entities |
| Certificate Signing Request | Request in order to apply for a digital identity certificate   |

## Motivation

A typical productive Kubernetes cluster could be hardened in many different ways, whereas probably many of these actions
would overlap and target similar weaknesses of a cluster.
For this version of the standard, the following points should be addressed:

* Kube-API rate limiting
* etcd compaction/defragmentation
* etcd backup
* Certificate Authority (CA) expiration avoidance

These robustness features should mainly increase the up-time of the Kubernetes cluster by avoiding downtimes either
because of internal problems or external threads like "Denial of Service" attacks.
Additionally, the etcd database should be strengthened with these features in order to provide a secure and robust
backend for the Kubernetes cluster.

## Design Considerations

In order to provide a conclusive standard, some design considerations need to be set beforehand:

### Kube-API rate limiting

> Rate limiting is the practice of preventing too many requests to the same server in some time frame. This can help prevent

Reviewer: ```suggestion
Rate limiting is the practice of preventing too many requests to the same server in some time frame. This can help to prevent
```

Author: Both sentences are correct and I would like to keep the original one, since it sounds better in this context (but it doesn't make a difference meaning wise)

could lead to slow responsiveness or even malfunctions.
In order to possibly mitigate some problems for the Kubernetes clusters, robustness features should be introduced into
the SCS standards. These would harden the cluster infrastructure against several problems, making failures less likely.

## Motivation

A typical productive Kubernetes cluster could be hardened in many different ways, whereas probably many of these actions
would overlap and target similar weaknesses of a cluster.
For this version of the standard, the following points should be addressed:

* Kube-API rate limiting
* etcd compaction/defragmentation
* etcd backup
* CA expiration avoidance

These robustness features should mainly increase the up-time of the Kubernetes cluster by avoiding downtimes either
because of internal problems or external threads like "Denial of Service" attacks.
Additionally, the ETCD database should be strengthened with these features in order to provide a secure and robust
backend for the Kubernetes cluster.

## Design Considerations

In order to provide a conclusive standard, some design considerations need to be set beforehand:

### Kube-API rate limiting

> Rate limiting is the practice of preventing too many simultaneous requests to the same server. This can help prevent

Reviewer 2: ```suggestion
Rate limiting is the practice of preventing too many requests to the same server in some time frame. This can help prevent
```

Author: Updated this part.

#### Flow control

Flow control for the Kubernetes API server can be provided by the API priority and fairness feature, which classifies
and isolates requests in a fine-grained way in order to prevent an overload of the API server.
The package introduces queues in order to not deny requests and dequeue them through Fair Queueing techniques.
Overall, the Flow control package introduces many different features like request queues, rule-based flow control,
different priority levels and rate limit maximums.
The concept documentation offers a more in-depth explanation of the feature:
[Flow Control](https://kubernetes.io/docs/concepts/cluster-administration/flow-control/)

### etcd compaction/defragmentation

etcd is a strongly consistent, distributed key-value store that provides a reliable way to store data that needs to be
accessed by a distributed system or cluster of machines. For these reasons, etcd was chosen as the default database
for Kubernetes.
In order to remain reliable, an etcd cluster needs periodic maintenance. This is necessary to maintain the etcd keyspace;
failure to do so could lead to a cluster-wide alarm, which would put the cluster into a limited-operation mode.

> To mitigate this scenario, the etcd keyspace can be compacted and defragmented.

Reviewer 2: I think defragmentation should be dropped from the document. AFAIU, the docs cited below do not indicate that periodic defragmentation would improve reliability (if anything, it disrupts operations for a period of time).

Already had a discussion about this [here](https://github.com/SovereignCloudStack/k8s-cluster-api-provider/pull/401#discussion_r1151471155).

Author: After reading the provided link and the docs @matofeder wrote, I understand where you're coming from.
I don't know if I would drop it completely, but instead make the whole defragmentation part optional.

I will rewrite the part and then we can maybe get a third opinion on that.

Reviewer 2: > I will rewrite the part and then we can maybe get a third opinion on that.

Thanks! Cannot argue with the new revised form :)

Still, if defragmentation is mentioned, I think that the trade-off's of periodic defragmentation should be clear.

AFAIU Pros:
- a cluster that ran out of space for some reason (probably due to subpar disk planning or some unmitigated malicious DoS actions) might "repair" itself with defragmentation at e. g. 2 AM

AFAIU Cons:
- Effort of maintaining defragmentation automation
- Disrupts client operations for a short time (especially if done on all members at the same time)
- If proper disk planning, client limits etc. are in pace, preventing out-of-space conditions, periodic defragmentation has little to no value

Reviewer 3: I would like to comment that even though @joshmue pointed out various drawbacks, I still find defragmentation to be valuable, my points:

-  If an infrastructure monitoring solution detects high disk utilization on the node housing etcd, I, as an OPS engineer, expect that performing maintenance on etcd should resolve the alert for high disk utilization.
 
- In addition, see https://github.com/etcd-io/etcd/issues/14267 which complains that compaction alone does not make free space usable. Also see the comment written by etcd maintainer: https://github.com/etcd-io/etcd/issues/14267#issuecomment-1193404153 

In SCS KaaS we did solid work on how to defragment etcd cluster to avoid cases like @joshmue pointed out: "Disrupts client operations for a short time (especially if done on all members at the same time)" 
- see related docs https://github.com/SovereignCloudStack/k8s-cluster-api-provider/blob/main/doc/Maintenance_and_Troubleshooting.md#etcd-maintenance

Reviewer 2: :+1: 

> `In addition, see Compaction does not make free space usable etcd-io/etcd#14267 which complains that compaction alone does not make free space usable. Also see the comment written by etcd maintainer: Compaction does not make free space usable etcd-io/etcd#14267 (comment)`

I'ts worth adding that this issue is concerned/complains about the documented behavior that defragmentation is required after running out of space - to make the cluster operable again. Hence, it is not strictly related to periodic defragmentation.
I actually do not know how to understand the maintainer's response regarding to usual operation (as well as the OP over there). The mentioned enhancement sounds more like a change in calculation of the space quota - which would block after an out-of-space condition right now (as documented), but (ideally) should not.

Author: > It's worth adding that this issue is concerned/complains about the documented behavior that defragmentation is required after running out of space - to make the cluster operable again. Hence, it is not strictly related to periodic defragmentation.

You're right that it's not strictly related to periodic defragmentation, but if the "free space" isn't usable after compaction without defragmentation, I would argue that compaction isn't necessarily that useful all together.
I would still prefer to do it, but it wouldn't have the same "value" as before.

What I would also like to know (regarding the previous comment), how exactly "disk planning" could help prevent out-of-space conditions. I guess forward planning is important in this regard, meaning that an etcd cluster should always have enough disk space to accompany the possible growth. I guess the question here would be how much "disk" a provider would be willing to put into a cluster as a kind of security so that no out-of-space would occur and what scenarios will be expected. From my experience in the field people won't have an eye on everything and can't prevent every problem before it happens. So I think such an automation would probably be a good thing.

I also added @matofeder recommendation about the avoidance of service disruption.

Reviewer 2: > You're right that it's not strictly related to periodic defragmentation, but if the "free space" isn't usable after compaction without defragmentation, I would argue that compaction isn't necessarily that useful all together.
> I would still prefer to do it, but it wouldn't have the same "value" as before.

Yes, compaction alone does not seem to help at all if the storage quota was reached and etcd does not allow any more writes. My point was just that this is a very different situation than running a cluster in normal operation, when compaction alone usually would prevent (combined with other limits or just adequately little load) the storage quota to be reached.

Maybe there are indeed cases where defragmentation really frees up space that would not be used otherwise - e. g. if there are many tiny gaps in the backend database, but all new data is bigger than any gap - like the OP of the mentioned issue theorized.
I know that the [OpenShift docs](https://docs.openshift.com/container-platform/4.14/scalability_and_performance/recommended-performance-scale-practices/recommended-etcd-practices.html#etcd-defrag_recommended-etcd-practices) strongly recommend it and implement it as automatic process, but I do not really understand why. Also seems to be [sometimes](https://access.redhat.com/solutions/6840041) [problematic](https://access.redhat.com/solutions/6960380) for such installations.

It's just that the documentation AFAIU does not hint that *periodic* defragmentation is of any advantage or recommended. Nor do I know of any POC or case (either by reading about it or by experience) where it was actually needed. The only plain statement I found was from a former etcd maintainer in https://github.com/etcd-io/etcd/issues/7607#issuecomment-289351605:

> This is actually about disk fragmentation. But if you do not suddenly remove a lot of keys and want to reclaim the disk space immediately, you do not need to defrag.

I'm not strictly opposed to adding docs how to do (periodic) defragmentation on etcd - just want everyone to have the same understanding on why or why not it should be mentioned/recommended. Also, if good reasons are found for implementing automatic defragmentation, the recommendation should also go into the etcd documentation.

---

> What I would also like to know (regarding the previous comment), how exactly "disk planning" could help prevent out-of-space conditions. I guess forward planning is important in this regard, meaning that an etcd cluster should always have enough disk space to accompany the possible growth. I guess the question here would be how much "disk" a provider would be willing to put into a cluster as a kind of security so that no out-of-space would occur and what scenarios will be expected. From my experience in the field people won't have an eye on everything and can't prevent every problem before it happens. So I think such an automation would probably be a good thing.

Again (AFAIU and under the assumption that etcd generally behaves as documented), defragmentation generally does not prevent out-of-space conditions.
In contrast: If disk planning ensures that etcd has always at least 2GB (or whatever storage size quota is configured) disk space available, the disk will not run out of space, as etcd hits its own quota first (AFAIU, assuming that the storage quota equals database file size).

Author: For now, I wrote the part about the defragmentation as completely optional into the `Decision` section.

Reviewer 2: Thanks for [adding](https://github.com/SovereignCloudStack/standards/compare/d8b01baa2aacc34c8f4eeabaa93d5a9d254b60de..e6583e9428e12a3aaddf1e4b99a0a80e2f49922c) the stability risks of defragmentation. :+1: 

> If thats not ok as it is, please make an edit to the text how your preferences would be for that section.

My only strong preference is really just to include *why* periodic defragmentation [^1] may be a good idea for cluster robustness and hence belong in this document about "K8s cluster robustness features".
If no reason is strong enough to be included in the document itself, I'd opt to remove the paragraphs about it completely (and maybe add a paragraph about guaranteeing disk space to etcd).

[^1]: As opposed to e. g. manual defragmentation as part of an operational handbook in case of an out-of-space condition

Author: I removed most of the defragmentation parts for now (except a very small mentioning in the `etcd maintenance` section).

I also decided to not add a paragraph about disk space guarantees for now.

* max-requests-inflight
* max-mutating-requests-inflight
* min-request-timeout

Values for these flags/parameters SHOULD be adapted to the needs of the environment and
the expected load.

A cluster MUST also activate and configure a Ratelimit admission controller.
This requires an `EventRateLimit` resource to be deployed on the Kubernetes cluster.
The following settings are RECOMMENDED for a cluster-wide deployment, but more
fine-grained rate limiting can also be applied, if this is necessary.

```yaml
kind: Configuration
apiVersion: eventratelimit.admission.k8s.io/v1alpha1
limits:
- burst: 20000
  qps: 5000
  type: Server
```

It is also RECOMMENDED to activate the Kubernetes API priority and fairness feature,
which also uses the aforementioned cluster parameters to better queue, schedule and
prioritize incoming requests.

### etcd compaction/defragmentation

etcd needs to be cleaned up regularly, so that it functions correctly and doesn't take
up too much space, which happens because of its increase of the keyspace.

To compact the etcd keyspace, the following flags/parameters MUST be set for etcd:

* auto-compaction-mode = periodic
* auto-compaction-retention = 8h

> A cluster defragmentation MUST be carried out regularly.

Author: See my answer above.

> A cluster defragmentation MUST be carried out regularly.

Reviewer 2: :+1: 

Resolving this thread in favor of the above one, then.

for Kubernetes.
In order to remain reliable, an etcd cluster needs periodic maintenance. This is necessary to maintain the etcd keyspace;
failure to do so could lead to a cluster-wide alarm, which would put the cluster into a limited-operation mode.
To mitigate this scenario, the etcd keyspace can be compacted. Additionally, an etcd cluster can be defragmented, which
gives back disk space to the underlying file system and can help bring the cluster back into an operable state, if it
ran out of space earlier.

This can be achieved by providing the necessary flags/parameters to etcd, either via the KubeadmControlPlane or in the
configuration file of the etcd cluster, if it is managed independent from the Kubernetes cluster.
Possible flags, that can be set for this feature, are:

* auto-compaction-mode
* auto-compaction-retention

etcd cluster defragmentation unfortunately can't be done automatically. Instead the user would need to manually call
the defrag command on the cluster. In order to mitigate this, a systemd (or similar) job could be created, which
periodically calls the defragmentation procedure.

> An example for such a job and its helpers could be the following:

Reviewer 3: Would be worth mentioning that defragmentation to a live etcd member blocks the system from reading and writing data while rebuilding its states. 

Alternatively, (my preferable option), would be great to recommend a defragmentation strategy as is described e.g. [here](https://docs.openshift.com/container-platform/4.9/scalability_and_performance/recommended-host-practices.html#automatic-defrag-etcd-data_recommended-host-practices) or also described and implemented [here](https://github.com/SovereignCloudStack/k8s-cluster-api-provider/blob/main/doc/Maintenance_and_Troubleshooting.md), 
i.e.:
- Defragment the non leader etcd members first
- Change the leadership to the defragmentation completed etcd member
- Defragment the last (ex-leader) etcd member

Author: I updated this section to include your recommendations. I already read them before and they make sense, so I don't really know why I didn't include them before.

For this reason, a backup strategy needs to be developed with a decreasing number of backups in an increasing period of time,
meaning that the previous year should only have 1 backup, but the current week should have multiple.
Information about the backup process can be found in the etcd documentation:
[Upgrade etcd](https://kubernetes.io/docs/tasks/administer-cluster/configure-upgrade-etcd/)

### CA expiration avoidance

In order to secure the communication of a Kubernetes cluster, (TLS) certificates signed by a controlled
Certificate Authority (CA) can be used.
Normally, these certificates expire after a set period of time. In order to avoid expiration and failure of a cluster,
these certificates need to be rotated regularly and at best automatically.
It is important to either set `--rotate-server-certificates` as a command line parameter or set `rotateCertificates: true`
in the kubelet config or the `kubeletExtraArgs` of the cluster-template.yaml file. This activates the rotation of the
kubelet server certificate. Another recommendation is to set `serverTLSBootstrap: true`, which also enables the request

> and rotation of the certificate for the kubelet according to the documentation

Reviewer 3: IMO I find the explanation regarding the rotation of kubelet server certificates somewhat unclear. In addition, according to the documentation, it appears that the approval of Certificate Signing Requests (CSRs) should be done either manually or by using a custom controller, as indicated in this note [here](https://kubernetes.io/docs/reference/access-authn-authz/kubelet-tls-bootstrapping/#certificate-rotation).

Overall, I do not see the recommendation to set the kubelet parameters to automate server certificate rotation as a complete solution because it is just a part of Kubernetes [PKI](https://kubernetes.io/docs/setup/best-practices/certificates/). 

Instead, I would prefer 2 approaches you described below, i.e. 

1. Update the cluster at least once a year to rotate certificates automatically
2.  Renew all certificates with `kubeadm certs renew all`
     - You only need to do this when you don't upgrade your cluster
     
the above is also in sync with SCS KaaS reference implementation, see [docs](https://github.com/SovereignCloudStack/k8s-cluster-api-provider/blob/main/doc/Maintenance_and_Troubleshooting.md#client-certificates-in-kubernetes-expire-after-one-year)

Author: I probably wrote this section a bit unclear then and I also probably understand something wrong in the certificate rotation documentation (and some other documents).
I planned this section to basically recommend the cluster renewal or certificate renewal approaches you described. But I was under the impression, that the flags described in the first subsection are required to have all the certificates exchanged correctly.

Reviewer 3: > But I was under the impression, that the flags described in the first subsection are required to have all the certificates exchanged correctly.

After rereading the docs (again), you are right. Kubeadm does not manage the kubelet cert rotation. 
Based on this [docs](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#kubelet-serving-certs), my understanding is as follows:

1. The command `kubeadm certs renew all` will renew Kubernetes control plane certificates, but it won't renew the certificates used for authenticating the kubelet against the k8s API server and won't renew the kubelet server certificate
2. The kubelet parameter `rotateCertificates` (defaults to true) will enable the full automatic renewal of kubelet client certificate for authenticating the kubelet against the k8s API
3.  The kubelet parameter  `TLSBootstrap: true` will enable the bootstrap of kubelet server certificates by requesting them from the certificates.k8s.io API. 
 ! One known limitation is that the CSRs (Certificate Signing Requests) for these certificates cannot be automatically approved. This will require manual action from the user or a third-party controller like [kubelet-csr-approver](https://github.com/postfinance/kubelet-csr-approver)

Author: I updated the section a bit more to include the possibility for a CSR approver controller and also a better representation for the `kubeadm certs renew all` command.

For this reason, a backup strategy needs to be developed with a decreasing number of backups in an increasing period of time,
meaning that the previous year should only have 1 backup, but the current week should have multiple.
Information about the backup process can be found in the etcd documentation:
[Upgrade etcd](https://kubernetes.io/docs/tasks/administer-cluster/configure-upgrade-etcd/)

### Certificate rotation

In order to secure the communication of a Kubernetes cluster, (TLS) certificates signed by a controlled
Certificate Authority (CA) can be used.
Normally, these certificates expire after a set period of time. In order to avoid expiration and failure of a cluster,
these certificates need to be rotated regularly and at best automatically.
Certificates can either be rotated manually (a reference for manually working with certificates can be found
[here](https://github.com/kelseyhightower/kubernetes-the-hard-way/blob/79a3f79b27bd28f82f071bb877a266c2e62ee506/docs/04-certificate-authority.md))
or automatically, which requires other things to care about in a deployment.

Some tools or clusters provide possibilities to rotate certificates manually.

> For example, `kubeadm` and `k3s` provides the following commands
> 
> ```bash
> # kubeadm
> kubeadm certs renew all
> 
> # k3s
> k3s certificate rotate

Reviewer 4: ```suggestion
For example, `kubeadm` provides the following command

```bash
# kubeadm
kubeadm certs renew all
```
Remove k3s to avoid confusion since we don't use this in KaaS

Author: I can do that, but I thought more about this a general example. Maybe "we" don't use it, but what if someone external wants to make his k3s cluster SCS conform?
Or do we generally say, that this isn't a thing?

Reviewer 4: @cah-hbaum ok let's keep it. Since there is a k3s providing Kubernetes on IaaS, it's probably not too far away from us.

Author: Dont know how to "resolve" your changes if we don't integrate them, maybe you need to approve or something similar.
